### PR TITLE
allow obj listeners with handleEvent methods Fixes #6

### LINKED ISSUE #6: listener functions don't HAVE to be functions
they could also be objects with the `handleEvent` method defined on their prototype chain.  [This check](https://github.com/mysticatea/event-target-shim/blob/master/lib/event-target.js#L111-L112) is preventing a consumer of this lib from passing an object to `addEventListener`.

### PR DESCRIPTION
**Pull request by nickdesaulniers**

### CODE REVIEW

**Comment by mysticatea** [comment]
Thank you!

Hmm, I guess `removeEventListener` for an object listener doesn't work in this implementation.

**Comment by nickdesaulniers** [comment]
oh, good catch.  I'll take a look.

**Comment by nickdesaulniers** [comment]
hmm...I'll have to think.

```js
function a () {}
b = a.bind(42)
c = a.bind(42)
a === b // false
```

Since in addEventListener, we're setting the listener to a bound function (`Function.prototype.bind` returns a new instance of the method, and thus a different reference), it's going to be hard to compare that reference to anything passed in to removeEventListener.

Maybe there's another way I'm missing?

**Comment by nickdesaulniers** [comment]
maybe it's better to avoid the call to bind and instead check for `typeof object` and handleEvent during dispatch?

**Comment by mysticatea** [comment]
I think we cannot use `Function.prototype.bind` there.

My idea is that `addEventListener` stores the `listener` as is, then `dispatchEvent` checks the `listener`.

```js
if (typeof node.listener === "function") {
    node.listener.call(this, wrapped);
}
else {
    node.listener.handleEvent(wrapped);
}
```

**Comment by nickdesaulniers** [comment]
check it out, think we're good to go now.  Let me know if you want the patches squashed or not.  About to catch a plane, though.

File: lib/event-target.js [hunk 1]
```diff
@@ -180,7 +186,14 @@ EventTarget.prototype = Object.create(
                 // This doesn't process capturing phase and bubbling phase.
                 // This isn't participating in a tree.
                 while (node != null) {
-                    node.listener.call(this, wrapped);
+                    var isObjListener = hasHandleEventMethod(node.listener);
+                    if (isObjListener) {
+                        node.listener.handleEvent.bind(node.listener).call(this, wrapped);
```
Comment (mysticatea) [inline lib/event-target.js]: `.bind(node.listener).call(this, ` is needless here.

```
node.listener.handleEvent(wrapped);
```

**Comment by mysticatea** [comment]
I noticed, `dispatchEvent` should not throw an error even after `delete obj.handleEvent`.

image

File: lib/event-target.js (hunk 1 above)
Comment (nickdesaulniers) [inline lib/event-target.js]: I tried removing it, the first test I added then fails since `this` is not correct.

Comment (nickdesaulniers) [inline lib/event-target.js]: Please verify.

**Comment by nickdesaulniers** [comment]
That's interesting, I would expect `isObjListener` to be false in that case.

File: lib/event-target.js (hunk 1 above)
Comment (mysticatea) [inline lib/event-target.js]: `node.listener.handleEvent(wrapped);` works in my PC :)

**Comment by mysticatea** [comment]
Hmm, OK.
I will merge this PR then tweak the behavior to adjust to well-known browsers.
Thank you for your good work!

**Comment by nickdesaulniers** [comment]
cool, thank you for accepting my PR, cleaning it up, and publishing it!  I appreciate it!

**Pull request merged**